ROX-18092: compliance scan config

dashrews78 opened the pull request:
## Description

The PR follows #7827.  This PR implements the APIs and storage of compliance scan configuration settings, pushes the request to sensor, and finally writes the necessary CRs within the compliance operator.

- Services implemented to:
   - post a scan configuration
   - retrieve a specific scan configuration
   - retrieve scan configurations matching query
   - convert functions to go between storage and API objects
- Datastores Implemented to:
   - Upsert a scan configuration
   - retrieve a specific scan configuration
   - delete a specific scan configuration
   - update the status of a scan configuration by cluster
   - retrieve the status by cluster of the scan configurations
- Manager implemented to:
   - Process request to create a scan configuration
       - validate scan configuration doesn't exist
       - validate the schedule
       - send the message to sensor
   - Handle any responses from sensor regarding the scan configuration
- Sensor updated to:
   -   fix a bug in creation of scansettings
   - handle `ComplianceResponse_ApplyComplianceScanConfigResponse_` response and send it to central
   - 

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
~- [ ] Evaluated and added CHANGELOG entry if required~
~- [ ] Determined and documented upgrade steps~
~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

### Here I tell how I validated my change

- Wrote unit tests

Tested locally:
- Created openshift cluster
- installed compliance operator
- Applied workaround to create role and role-bindings referenced in https://issues.redhat.com/browse/ROX-18124
- Ran `roxcurl /v2/compliance/scan/configurations -d '{"scanName":"shrewstest","clusters":["CLUSTERID"],"scanConfig":{"profiles":["ocp4-cis"],"oneTimeScan":false}}' -u admin:PASSHERE` to post a scan config that uses the default schedule
- Ran `roxcurl /v2/compliance/scan/configurations -d '{"scanName":"shrewstest11","clusters":["CLUSTERID"],"scanConfig":{"profiles":["rhcos4-moderate"],"oneTimeScan":false,"scanSchedule":{"intervalType":"WEEKLY","hour":0,"minute":0,"daysOfWeek":{"days":[0,1,2,3,4,5,6]}}}}' -u admin:PASSHERE` to run different profile at midnight every day
- Ran `roxcurl /v2/compliance/scan/configurations -d '{"scanName":"shrewstest9","clusters":["CLUSTERID"],"scanConfig":{"profiles":["rhcos4-moderate"],"oneTimeScan":false,"scanSchedule":{"intervalType":1,"hour":0,"minute":0,"interval":{"daysOfWeek":{"days":[0,1,2,3,4,5,6]}}}}}' -u admin:PASSHERE` to ensure an error was quickly returned due to invalid schedule.
- used oc commands to verify scan settings were written and scans executed
    - oc get -n openshift-compliance scansettings
    - oc get -n openshift-compliance compliancesuites
    - oc get -n openshift-compliance compliancecheckresults -l compliance.openshift.io/suite=shrewstest 

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

Test status is weird.  Tests like the profiling ones haven't been working for anyone and thus unrelated.  `ci/prow/ocp-4-13-ebpf-qa-e2e-tests` has been failing widely with collector startup issues.  `ci/prow/ocp-4-13-qa-e2e-tests` actually passed but status was not reported back to the PR.  `ci/prow/ocp-4-10-qa-e2e-tests` was environmental because it couldn't find an ENI when starting the cluster.

dashrews78 committed:
X-Smart-Branch-Parent: master

md2119 commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -3,32 +3,65 @@ package compliancemanager
-	integrationDS integrationDS.DataStore
+	integrationDS compIntegration.DataStore
+	scanSettingDS compScanSetting.DataStore
+	requests      map[string]clusterScan
```
Is the key `requestID` which has 1:n mapping to scans?
Does this hold currently running requests or just a cache without invariants?

dashrews78 replied:
@md2119 
This only holds currently running requests.

I started down that path of using one requestID for all clusters but dealing with the 1:n seemed unnecessarily complicated.  Down below I generate a unique id to pass into each sensor message.  Obviously I should probably name that variable better.  The idea is this will hold the requests that are out to the sensors so that we can easily update the status when the message comes back.  

```
	for _, cluster := range clusters {
		// id for the request message to sensor
		id := uuid.NewV4().String()
```
Now that I'm explaining this, a lot of it is a remnant when I had the status buried in the configuration object and was trying to be surgical and minimize updates to the configuration object.  Since I split out the cluster status as a separate table for state, I could probably use the table instead as it is per cluster with a FK back to the configuration.

Opinions?

md2119 replied:
Thanks for the explanation. A couple of comments to remove the confusion:
- The intent can be clarified by renaming the field to `openRequests`/`runningRequests`
- A comment to provide clarification about the field. It is easy to confuse if the key is the ID of the request that the user submitted (which is what thought to it to be, hence 1:n) or the ID of the request sent by Central to Sensors.

With the above in mind, each communication loop, could be a single roundtrip or more, between Central and Sensor should have a unique ID.

md2119 commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -3,32 +3,65 @@ package compliancemanager
 // ProcessComplianceOperatorInfo processes and stores the compliance operator metadata coming from sensor
 func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complianceIntegration *storage.ComplianceIntegration) error {
+	if !features.ComplianceEnhancements.Enabled() {
+		return nil
```
Better to return error to give users the clue.

md2119 commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +87,152 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+	if scanRequest.GetSchedule() != nil {
+		cron, err = schedule.ConvertToCronTab(scanRequest.GetSchedule())
+		if err != nil {
+			log.Error(err)
+			return nil, errors.Errorf("Unable to convert schedule for scan configuration named %q to cron.", scanRequest.GetScanName())
```
The `err` from conversion while include insights to fix the problem. WIthout that, it could be hard for the users to figure out the problem.

```suggestion
			err =  errors.Wrapf(err, "Unable to convert schedule for scan configuration named %q to cron.", scanRequest.GetScanName())
			log.Error(err)
			return nil, err
```

md2119 commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +87,152 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+		cronValidator := gronx.New()
+		if !cronValidator.IsValid(cron) {
+			log.Error(err)
+			return nil, errors.Errorf("Schedule for scan configuration named %q is invalid.", scanRequest.GetScanName())
```
Same as above for such instances.

md2119 commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +87,152 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+		profiles = append(profiles, profile.GetProfileName())
+	}
+
+	m.requestsLock.Lock()
+	defer m.requestsLock.Unlock()
```
You should not hold lock around a grpc call. You only need to protect the updates the `requests` map.

md2119 commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +87,152 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+			status = err.Error()
+			log.Errorf("error sending compliance scan config to cluster %q: %v", cluster, err)
+		} else {
+			// Request was not rejected so add it to map awaiting response
+			m.requests[id] = clusterScan{
+				clusterID: cluster,
+				scanID:    scanRequest.GetId(),
+			}
```
move this with locking to a separate func.

md2119 commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +87,152 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+	m.requestsLock.Lock()
+	defer m.requestsLock.Unlock()
+
+	var scanID string
+	if clusterScanData, found := m.requests[requestID]; found {
+		if clusterScanData.clusterID != clusterID {
+			return errors.Errorf("Cluster mismatch for request %q", requestID)
+		}
+		scanID = clusterScanData.scanID
+	} else {
+		return errors.Errorf("Unable to find request %q", requestID)
+	}
```
This does not take into consideration that Central can restart after sending the request and before receiving the response.

dashrews78 replied:
@md2119 Right.  I figured we would take care of that with the sync of configurations under https://issues.redhat.com/browse/ROX-18711.  At that time we would get the latest status of any configurations on the clusters, and apply any configurations that need it.  Does that sound reasonable?  I should probably put in a comment about that.

md2119 replied:
Maybe add a TODO to this code block

md2119 reviewed (changes requested):
I reviewed the PR partially. My biggest concern is the improper `requests` map management; refer to comment.

md2119 commented on central/complianceoperator/v2/scanconfigurations/datastore/datastore.go:
```diff
@@ -0,0 +1,72 @@
+	// UpsertScanConfiguration adds or updates the scan configuration
+	UpsertScanConfiguration(ctx context.Context, scanConfig *storage.ComplianceOperatorScanConfigurationV2) error
+
+	// DeleteScanConfiguration deletes the scan configuration specified by name
```
`name` -> `id`

md2119 commented on central/complianceoperator/v2/scanconfigurations/datastore/datastore.go:
```diff
@@ -0,0 +1,72 @@
+	// UpdateClusterStatus updates the scan configuration with the cluster status
+	UpdateClusterStatus(ctx context.Context, scanID string, clusterID string, clusterStatus string) error
+
+	// GetScanConfigClusterStatus retrieves the scan configurations status per cluster specified by scan name
```
`name` to `ID`

md2119 commented on central/complianceoperator/v2/scanconfigurations/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,135 @@
+	return ds.storage.Upsert(ctx, scanConfig)
+}
+
+// DeleteScanConfiguration deletes the scan configuration specified by name
```
id

md2119 commented on central/complianceoperator/v2/scanconfigurations/service/convert.go:
```diff
@@ -0,0 +1,235 @@
+
+	storageIntervalTypeToV2 = map[storage.Schedule_IntervalType]v2.Schedule_IntervalType{
+		storage.Schedule_UNSET:   v2.Schedule_UNSET,
+		storage.Schedule_DAILY:   v2.Schedule_UNSET,
```
Is daily not supported at API?

dashrews78 replied:
I was able to add Daily support with no trouble.

dashrews78 committed:
X-Smart-Squash: Squashed 21 commits:

39b43b266b small refactor and some minor proto adjustments
db0939e00b clean up
5230c9655c wire up config
bab0d40e8f test fixes
e2f0467137 missed a file
03ae4cde45 log cron
cdf5f51cff schedule converter fix
57bae8f232 validate schedule before it goes to sensor
8fd9153454 push error back
52011429b8 some clean up
852b83b573 fix UT
b9e40b8232 some clean up
83b8dbf7cf some clean up
deb19c7340 finish a rebase
0ee3b0bfee style and test
8216d8a055 missed nil in schedule conversion
9cc2493053 addressing feedback
954dab45b6 add a comment
70e6218902 add daily schedule
e6d4824d48 addressing some comments on map
4e0c695bbd fix creating schedule error

md2119 commented on central/complianceoperator/v2/scanconfigurations/service/service_impl.go:
```diff
@@ -0,0 +1,136 @@
+	return ctx, authorizer.Authorized(ctx, fullMethodName)
+}
+
+func (s *serviceImpl) PostComplianceScanConfiguration(ctx context.Context, req *v2.ComplianceScanConfiguration) (*v2.ComplianceScanConfiguration, error) {
```
`Post...` -> `Create...`

For more details: https://github.com/stackrox/stackrox/blob/master/.github/api_guidelines.md

md2119 commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +90,162 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+	}
+
+	// Check if scan configuration already exists.
+	found, err := m.scanSettingDS.GetScanConfigurationExists(ctx, scanRequest.GetScanName())
```
What was the motivation to choose `GetScanConfigurationExists` over `ScanConfigurationExists`?

dashrews78 replied:
no motivation.  I copied a method signature and stuck exists on the end in a rush.  I will update it

dashrews78 commented on central/complianceoperator/v2/scanconfigurations/service/convert.go:
```diff
@@ -0,0 +1,236 @@
+		ClusterStatus: func() []*v2.ClusterScanStatus {
+			clusterStatuses := make([]*v2.ClusterScanStatus, 0, len(scanClusters))
+			for _, cluster := range scanClusters {
+				clusterName, found, err := clusterStore.GetClusterName(ctx, cluster.GetClusterId())
+				if err != nil {
+					log.Error(err)
+				}
+				if !found {
+					log.Errorf("unable to find cluster name for cluster %q", cluster.GetClusterId())
+				}
+				clusterStatuses = append(clusterStatuses, &v2.ClusterScanStatus{
+					ClusterId:   cluster.GetClusterId(),
+					ClusterName: clusterName,
+					Errors:      cluster.GetErrors(),
+				})
```
yeah I was trying to avoid storing it, but that seems to make it more complicated than need be.  I will store it.

dashrews78 committed:
store cluster name in config status

dashrews78 committed:
test fixes

dashrews78 committed:
missed rename in mock usage

dashrews78 committed:
make sure cluster name is populated

dashrews78 commented:
/retest

rhmdnd commented on sensor/kubernetes/complianceoperator/utils.go:
```diff
@@ -45,7 +45,7 @@ func convertCentralRequestToScanSettingBinding(namespace string, request *centra
 	for _, profile := range request.GetProfiles() {
 		profileRefs = append(profileRefs, v1alpha1.NamedObjectReference{
 			Name:     profile,
-			Kind:     complianceoperator.ScanSetting.Kind,
```
Interesting - this must have been causing an issue during integration tests?

dashrews78 replied:
Yes that was causing an error when writing or running the objects.  If memory serves it could not find the profile until I updated this.

rhmdnd replied:
Got it - yeah that makes sense. Thanks!

rhmdnd commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +90,165 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+	// Convert and validate schedule
+	// TODO(ROX-19818):  remove default schedule
+	// For MVP a schedule is required.  If one is not present use daily at midnight.  As it evolves
+	// it may be as simple as no schedule means a one time scan.
```
++

That makes sense

md2119 reviewed (approved):
LGTM. 

I would recommend to keep the SAC simple. Therefore, instead of having individual perm for scan configuration, profiles, etc, use only one SAC perm. The write permission gives the users access to perform create, update, and delete operations in compliance, whereas the read gives them read-only access to compliance.

rhmdnd commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +90,165 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+	err = m.scanSettingDS.UpsertScanConfiguration(ctx, scanRequest)
+	if err != nil {
+		log.Error(err)
+		return nil, errors.Errorf("Unable to save scan configuration named %q.", scanRequest.GetScanName())
```
Do we typically refrain from bubbling up errors from the backend? I see we're returning the error from sensor below, but not here, and just wondering if there is a reason for the discrepancy.

dashrews78 replied:
That depends.  In this case that would be a SQL error and the error would be pushed back to the service and user.  We have generally tried to not push those back to the user.

rhmdnd commented on central/complianceoperator/v2/compliancemanager/manager_impl.go:
```diff
@@ -54,23 +90,165 @@ func (m *managerImpl) ProcessComplianceOperatorInfo(ctx context.Context, complia
+// HandleScanRequestResponse processes response of compliance scan configuration from a sensor.
+func (m *managerImpl) HandleScanRequestResponse(ctx context.Context, requestID string, clusterID string, responsePayload string) error {
+	if !features.ComplianceEnhancements.Enabled() {
+		return nil
```
We swallow this because it's ultimately coming from sensor and not an end-user directly?

dashrews78 replied:
yeah, that may be useful.

dashrews78 committed:
some minor rework suggestions

dashrews78 commented:
> LGTM.
> 
> I would recommend to keep the SAC simple. Therefore, instead of having individual perm for scan configuration, profiles, etc, use only one SAC perm. The write permission gives the users access to perform create, update, and delete operations in compliance, whereas the read gives them read-only access to compliance.

Agreed on SAC.  Right now I'm going to continue re-using the `Compliance Operator` resource until ROX-19742.  But that is global scope.  We may want to narrow the scope to cluster when we implement that ticket.

dashrews78 commented:
/retest
